internal/scheduler/nodedb/nodedb.go

```diff
@@ -523,60 +541,89 @@ func BindJobToNode(priorityClasses map[string]configuration.PriorityClass, job i
-	jobId := job.GetId()
-	queue := job.GetQueue()
-	requests := job.GetResourceRequirements().Requests
-
+// Specifically:
+//
+//   - The jobs that jobFilter returns true for are marked as evicted on the node.
+//   - Within AllocatableByPriorityAndResource, the resources allocated to these jobs are moved from
+//     the jobs' priorities to evictedPriority; they are not subtracted from AllocatedByJobId and
+//     AllocatedByQueue.
+func EvictJobsFromNode(
+	priorityClasses map[string]configuration.PriorityClass,
+	jobFilter func(interfaces.LegacySchedulerJob) bool,
+	jobs []interfaces.LegacySchedulerJob,
+	node *schedulerobjects.Node,
+) ([]interfaces.LegacySchedulerJob, *schedulerobjects.Node, error) {
+	evicted := make([]interfaces.LegacySchedulerJob, 0)
 	node = node.DeepCopy()
-
-	// Ensure we track allocated resources at evictedPriority.
-	if _, ok := node.AllocatableByPriorityAndResource[evictedPriority]; !ok {
-		pMin := int32(math.MaxInt32)
-		ok := false
-		for p := range node.AllocatableByPriorityAndResource {
-			if p < pMin {
-				pMin = p
-				ok = true
-			}
+	if jobFilter == nil {
+		return evicted, node, nil
+	}
+	for _, job := range jobs {
+		if !jobFilter(job) {
+			continue
 		}
-		if ok {
-			node.AllocatableByPriorityAndResource[evictedPriority] = node.AllocatableByPriorityAndResource[pMin].DeepCopy()
+		evicted = append(evicted, job)
+		if err := evictJobFromNodeInPlace(priorityClasses, job, node); err != nil {
+			return nil, nil, err
 		}
 	}
+	return evicted, node, nil
+}
 
+// evictJobFromNodeInPlace is the in-place operation backing EvictJobsFromNode.
+func evictJobFromNodeInPlace(priorityClasses map[string]configuration.PriorityClass, job interfaces.LegacySchedulerJob, node *schedulerobjects.Node) error {
+	jobId := job.GetId()
 	if _, ok := node.AllocatedByJobId[jobId]; !ok {
-		return nil, errors.Errorf("job %s has no resources allocated on node %s", jobId, node.Id)
+		return errors.Errorf("job %s has no resources allocated on node %s", jobId, node.Id)
 	}
+
+	queue := job.GetQueue()
 	if _, ok := node.AllocatedByQueue[queue]; !ok {
-		return nil, errors.Errorf("queue %s has no resources allocated on node %s", queue, node.Id)
+		return errors.Errorf("queue %s has no resources allocated on node %s", queue, node.Id)
 	}
+
 	if node.EvictedJobRunIds == nil {
 		node.EvictedJobRunIds = make(map[string]bool)
 	}
 	if _, ok := node.EvictedJobRunIds[jobId]; ok {
-		// TODO: We're using run ids instead of job ids for now.
-		return nil, errors.Errorf("job %s is already evicted from node %s", jobId, node.Id)
-	} else {
-		node.EvictedJobRunIds[jobId] = true
+		return errors.Errorf("job %s is already evicted from node %s", jobId, node.Id)
 	}
+	node.EvictedJobRunIds[jobId] = true
 
+	if _, ok := node.AllocatableByPriorityAndResource[evictedPriority]; !ok {
```

Comment:
We can get rid of this check once we stop storing `Node` values in the node database.